## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-zoned.sh | :x: [more...](#8dde015f-f43a-4a74-97aa-71b5f0ef5e9a) |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#331abacc-e6a0-43c8-9442-b28acb392a1a) |
<a name="331abacc-e6a0-43c8-9442-b28acb392a1a"></a>
### 12208-7530f62/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:31859
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 4376937f5b7a:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 3 seconds
(5/15) Running awscli tests ... FAILED in 32 seconds
{
  "name": "awscli",
  "duration": 2759,
  "function": "aws --endpoint-url http://minio-c3.minio.io:31859 s3api copy-object --bucket awscli-mint-test-bucket-13220 --key datafile-1-kB-copy --copy-source awscli-mint-test-bucket-13220/datafile-1-kB\n",
  "status": "FAIL",
  "error": "Hash mismatch expected 084e1383b70fb0c51acc680fef370023, got ac57de7156d7fc25ac1a65f81fa3989b"
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 48 seconds
(7/15) Running minio-dotnet tests ... done in 49 seconds
(8/15) Running minio-go tests ... FAILED in 2 minutes and 12 seconds
{
  "args": {},
  "duration": 568,
  "error": "At least one of the pre-conditions you specified did not hold",
  "function": "CopyObjectPart(destination, source)",
  "message": "CopyObjectPart call failed",
  "name": "minio-go: testUnencryptedToSSES3CopyObjectPart",
  "status": "FAIL"
}
(8/15) Running minio-java tests ... FAILED in 1 minutes and 50 seconds
{
  "name": "minio-java",
  "function": "copyObject()",
  "args": "[match etag]",
  "duration": 102,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = PreconditionFailed, message = At least one of the pre-conditions you specified did not hold, bucketName = minio-java-test-3e9tq2h, objectName = minio-java-test-2asosqe-copy, resource = /minio-java-test-3e9tq2h/minio-java-test-2asosqe-copy, requestId = 167BAEC15F329BC0, hostId = 94f14044-6cc8-402c-8ede-cdae2b319fbb)\nrequest={method=PUT, url=http://minio-c3.minio.io:31859/minio-java-test-3e9tq2h/minio-java-test-2asosqe-copy, headers=x-amz-copy-source-if-match: 71cff0a060f852067e443ad1e24ae26c-1\nx-amz-copy-source: /minio-java-test-pk93kj/minio-java-test-2asosqe\nHost: minio-c3.minio.io:31859\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210503T222356Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210503/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-copy-source;x-amz-copy-source-if-match;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=412, headers=Accept-Ranges: bytes\nContent-Length: 418\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nETag: \"71cff0a060f852067e443ad1e24ae26c\"\nLast-Modified: Mon, 03 May 2021 22:23:56 GMT\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 167BAEC15F329BC0\nX-Xss-Protection: 1; mode=block\nDate: Mon, 03 May 2021 22:23:56 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.execute(MinioClient.java:563), io.minio.MinioClient.executePut(MinioClient.java:904), io.minio.MinioClient.copyObject(MinioClient.java:1232), FunctionalTest.testCopyObjectMatchETag(FunctionalTest.java:1850), FunctionalTest.copyObject(FunctionalTest.java:2016), FunctionalTest.runObjectTests(FunctionalTest.java:3757), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(8/15) Running minio-js tests ... done in 46 seconds
(9/15) Running minio-py tests ... done in 2 minutes and 53 seconds
(10/15) Running s3cmd tests ... FAILED in 5 seconds
{
  "name": "s3cmd",
  "duration": "2983",
  "function": "test_put_object_multipart",
  "status": "FAIL",
  "error": "WARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Too many failures. Giving up on '/mint/data/datafile-65-MB'\nERROR: \nUpload of '/mint/data/datafile-65-MB' part 1 failed. Use\n  /usr/local/bin/s3cmd abortmp s3://s3cmd-test-bucket-14461/s3cmd-test-object-1654 3f923a27-46f8-4466-aac5-28549b049a05\nto abort the upload, or\n  /usr/local/bin/s3cmd --upload-id 3f923a27-46f8-4466-aac5-28549b049a05 put ...\nto continue the upload.\nERROR: Upload of '/mint/data/datafile-65-MB' failed too many times (Last reason: )"
}
(10/15) Running s3select tests ... done in 6 seconds
(11/15) Running security tests ... done in 0 seconds

Executed 11 out of 15 tests successfully.
```
<a name="8dde015f-f43a-4a74-97aa-71b5f0ef5e9a"></a>
### 12208-7530f62/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:31235
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 1c0584d18c71:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 3 seconds
(5/15) Running awscli tests ... done in 1 minutes and 28 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 47 seconds
(8/15) Running minio-dotnet tests ... done in 43 seconds
(9/15) Running minio-go tests ... done in 2 minutes and 21 seconds
(10/15) Running minio-java tests ... done in 1 minutes and 49 seconds
(11/15) Running minio-js tests ... done in 49 seconds
(12/15) Running minio-py tests ... FAILED in 2 minutes and 36 seconds
{
  "name": "minio-py:test_thread_safe",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-39530eca-04d1-4a89-90d1-45f764106bf6",
    "object_name": "e04183e8-1817-4ee0-add3-fc0a2fd4204f"
  },
  "message": "S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-39530eca-04d1-4a89-90d1-45f764106bf6, request_id: 167BAEFE88D2B37D, host_id: b70496b5-84e1-42b0-92f5-4d6dc8d70ff5, bucket_name: minio-py-test-39530eca-04d1-4a89-90d1-45f764106bf6",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 1640, in test_thread_safe\n    _CLIENT.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 676, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-39530eca-04d1-4a89-90d1-45f764106bf6, request_id: 167BAEFE88D2B37D, host_id: b70496b5-84e1-42b0-92f5-4d6dc8d70ff5, bucket_name: minio-py-test-39530eca-04d1-4a89-90d1-45f764106bf6\n",
  "duration": 6907
}
(12/15) Running s3cmd tests ... done in 18 seconds
(13/15) Running s3select tests ... done in 5 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/bitrot-streaming.go
+++ b/cmd/bitrot-streaming.go
@@ -25,6 +25,7 @@ import (
 	"hash"
 	"io"
 
+	xhttp "github.com/minio/minio/cmd/http"
 	"github.com/minio/minio/cmd/logger"
 	"github.com/minio/minio/pkg/ioutil"
 )
@@ -118,6 +119,14 @@ func (b *streamingBitrotReader) Close() error {
 		return nil
 	}
 	if closer, ok := b.rc.(io.Closer); ok {
+		// drain the body for connection re-use at network layer.
+		xhttp.DrainBody(struct {
+			io.Reader
+			io.Closer
+		}{
+			Reader: b.rc,
+			Closer: closeWrapper(func() error { return nil }),
+		})
 		return closer.Close()
 	}
 	return nil

--- a/cmd/erasure-heal_test.go
+++ b/cmd/erasure-heal_test.go
@@ -24,6 +24,9 @@ import (
 	"io"
 	"os"
 	"testing"
+
+	humanize "github.com/dustin/go-humanize"
+	"github.com/minio/minio/pkg/bpool"
 )
 
 var erasureHealTests = []struct {
@@ -136,8 +139,15 @@ func TestErasureHeal(t *testing.T) {
 				erasure.ShardFileSize(test.size), test.algorithm, erasure.ShardSize(), true)
 		}
 
+		// Number of buffers, max 2GB
+		n := (2 * humanize.GiByte) / (int(test.blocksize) * 2)
+
+		// Initialize byte pool once for all sets, bpool size is set to
+		// setCount * setDriveCount with each memory upto blockSizeV2.
+		bp := bpool.NewBytePoolCap(n, int(test.blocksize), int(test.blocksize)*2)
+
 		// test case setup is complete - now call Heal()
-		err = erasure.Heal(context.Background(), readers, staleWriters, test.size)
+		err = erasure.Heal(context.Background(), readers, staleWriters, test.size, bp)
 		closeBitrotReaders(readers)
 		closeBitrotWriters(staleWriters)
 		if err != nil && !test.shouldFail {

--- a/cmd/erasure-healing.go
+++ b/cmd/erasure-healing.go
@@ -451,7 +451,7 @@ func (er erasureObjects) healObject(ctx context.Context, bucket string, object s
 						tillOffset, DefaultBitrotAlgorithm, erasure.ShardSize(), true)
 				}
 			}
-			err = erasure.Heal(ctx, readers, writers, partSize)
+			err = erasure.Heal(ctx, readers, writers, partSize, er.bp)
 			closeBitrotReaders(readers)
 			closeBitrotWriters(writers)
 			if err != nil {

--- a/cmd/erasure-lowlevel-heal.go
+++ b/cmd/erasure-lowlevel-heal.go
@@ -20,30 +20,44 @@ package cmd
 import (
 	"context"
 	"io"
+	"sync"
 
 	"github.com/minio/minio/cmd/logger"
+	"github.com/minio/minio/pkg/bpool"
 )
 
 // Heal heals the shard files on non-nil writers. Note that the quorum passed is 1
 // as healing should continue even if it has been successful healing only one shard file.
-func (e Erasure) Heal(ctx context.Context, readers []io.ReaderAt, writers []io.Writer, size int64) error {
+func (e Erasure) Heal(ctx context.Context, readers []io.ReaderAt, writers []io.Writer, size int64, bp *bpool.BytePoolCap) error {
 	r, w := io.Pipe()
+	var wg sync.WaitGroup
+	wg.Add(1)
 	go func() {
-		if _, err := e.Decode(ctx, w, readers, 0, size, size, nil); err != nil {
-			w.CloseWithError(err)
-			return
-		}
-		w.Close()
+		defer wg.Done()
+		_, err := e.Decode(ctx, w, readers, 0, size, size, nil)
+		w.CloseWithError(err)
 	}()
-	buf := make([]byte, e.blockSize)
+
+	// Fetch buffer for I/O, returns from the pool if not allocates a new one and returns.
+	var buffer []byte
+	switch {
+	case size == 0:
+		buffer = make([]byte, 1) // Allocate atleast a byte to reach EOF
+	case size >= e.blockSize:
+		buffer = bp.Get()
+		defer bp.Put(buffer)
+	case size < e.blockSize:
+		// No need to allocate fully blockSizeV1 buffer if the incoming data is smaller.
+		buffer = make([]byte, size, 2*size+int64(e.parityBlocks+e.dataBlocks-1))
+	}
+
 	// quorum is 1 because CreateFile should continue writing as long as we are writing to even 1 disk.
-	n, err := e.Encode(ctx, r, writers, buf, 1)
-	if err != nil {
-		return err
-	}
-	if n != size {
+	n, err := e.Encode(ctx, r, writers, buffer, 1)
+	if err == nil && n != size {
 		logger.LogIf(ctx, errLessData)
-		return errLessData
+		err = errLessData
 	}
-	return nil
+	r.CloseWithError(err)
+	wg.Wait()
+	return err
 }

